src/main/java/baritone/process/BuilderProcess.java

```diff
@@ -344,13 +346,13 @@ private Optional<Placement> searchForPlacables(BuilderCalculationContext bcc, Li
                         }
                         desirableOnHotbar.add(desired);
                         Optional<Placement> opt = possibleToPlace(desired, x, y, z, bcc.bsi);
-                        if (opt.isPresent()) {
-                            return opt;
-                        }
+                        opt.ifPresent(list::add);
```

Comment:
Can we have an early return here once the list size reaches the placement limit?

Reply:
> Can we have an early return here once the list size reaches the placement limit?

I do that at line 571, The reason you don't want to do it here is I believe checks are preformed that shorten the list.

Reply:
Where are those additional checks? As far as I can tell it attempts precisely the first `placementLimit` results.

(Unrelated: I don't think `Optional` is still needed. The list can just be empty.)

Reply:
> Where are those additional checks? As far as I can tell it attempts precisely the first `placementLimit` results.

Line 348, instead of returning the first valid placement it returns upto the limit's amount of valid placements.

> (Unrelated: I don't think `Optional` is still needed. The list can just be empty.)
Also I believe a empty list is a code smell, and it ends

Reply:
Unless I missed something it returns *all* valid placements and then executes the first `placementLimit` ones, without any further checks. So any placement found after the first `placementLimit` ones will always be wasted time.

> Also I believe a empty list is a code smell

Lol, ok.

Reply:
> Unless I missed something it returns _all_ valid placements and then executes the first `placementLimit` ones, without any further checks. So any placement found after the first `placementLimit` ones will always be wasted time.

Well thats why I have the code at line 570.

Reply:
Yeah fine I changed it.